[MLIR][NFC] Stop depending on func.func in affine LoopUtils via &Region

last5bits opened the pull request:
Note: this is an experiment illustrating an approach alternative to PR https://github.com/llvm/llvm-project/pull/82079

Instead, pass a reference to the outermost region, which makes it possible to use affine LoopUtils routines in downstream dialects that have their own function-like ops.

joker-eph commented on mlir/lib/Dialect/Affine/Transforms/AffineDataCopyGeneration.cpp:
```diff
@@ -116,6 +116,8 @@ void AffineDataCopyGeneration::runOnBlock(Block *block,
                                    fastMemorySpace, tagMemorySpace,
                                    fastMemCapacityBytes};
+  auto &topRegion =
+      block->getParent()->getParentOfType<func::FuncOp>().getBody();
```
`getParentOfType` can fail which needs to be handled (fallback to getParentRegion?), and we also should check for a FunctionOpInterface instead for generality.

(here and everywhere else)

joker-eph commented on mlir/include/mlir/Dialect/Affine/LoopUtils.h:
```diff
@@ -273,7 +277,7 @@ void mapLoopToProcessorIds(scf::ForOp forOp, ArrayRef<Value> processorId,
+void gatherLoops(Region &region,
                  std::vector<SmallVector<AffineForOp, 2>> &depthToLoops);
 
 /// Creates an AffineForOp while ensuring that the lower and upper bounds are
```
We need to document "topRegion" everywhere

joker-eph commented:
Looks like a good direction I think

last5bits committed:
[MLIR][NFC] Stop depending on func.func in affine LoopUtils via &Region

Instead, pass a reference to the outermost region, which makes it
possible to use affine LoopUtils routines in downstream dialects that
have their own function-like ops.
